lib/ascii_binder/engine.rb

```diff
@@ -0,0 +1,872 @@
+      end.uniq
+    end
+
+    def validate_distros distros_string
+      return false if not distros_string.is_a?(String)
+      values = parse_distros(distros_string, true)
+      values.each do |v|
+        return false if not v == 'all' and not distro_map.distro_keys.include?(v)
+      end
+      return true
+    end
+
+    def validate_topic_group group, info
+      # Check for presence of topic group keys
+      ['Name','Dir','Topics'].each do |group_key|
+        if not group.has_key?(group_key)
+          raise "One of the topic groups in #{build_config_file} is missing the '#{group_key}' key."
+        end
+      end
+      # Check for right format of topic group values
+      ['Name','Dir'].each do |group_key|
+        if [true, false].include?(group[group_key])
+          raise "One of the topic groups in #{build_config_file} is using a reserved YAML keyword for the #{group_key} setting. In order to prevent your text from being turned into a true/false value, wrap it in quotes."
+        end
+        if not group[group_key].kind_of?(String)
+          raise "One of the topic groups in #{build_config_file} is not using a string for the #{group_key} setting; current value is #{group[group_key].inspect}"
+        end
+        if group[group_key].empty? or group[group_key].match BLANK_STRING_RE
+          raise "One of the topic groups in #{build_config_file} is using a blank value for the #{group_key} setting."
+        end
+      end
+      if not File.exists?(File.join(source_dir,info[:path]))
+        raise "In #{build_config_file}, the directory path '#{info[:path]}' for topic group #{group['Name']} does not exist under #{source_dir}"
+      end
+      # Validate the Distros setting
+      if group.has_key?('Distros')
+        if not validate_distros(group['Distros'])
+          key_list = distro_map.distro_keys.map{ |k| "'#{k}'" }.sort.join(', ')
+          raise "In #{build_config_file}, the Distros value #{group['Distros'].inspect} for topic group #{group['Name']} is not valid. Legal values are 'all', #{key_list}, or a comma-separated list of legal values."
+        end
+        group['Distros'] = parse_distros(group['Distros'])
+      else
+        group['Distros'] = parse_distros('all')
+      end
+      if not group['Topics'].is_a?(Array)
+        raise "The #{group['Name']} topic group in #{build_config_file} is malformed; the build system is expecting an array of 'Topic' definitions."
+      end
+      # Generate an ID for this topic group
+      group['ID'] = camelize group['Name']
+      if info.has_key?(:parent_id)
+        group['ID'] = "#{info[:parent_id]}::#{group['ID']}"
+      end
+    end
+
+    def validate_topic_item item, info
+      ['Name','File'].each do |topic_key|
+        if not item[topic_key].is_a?(String)
+          raise "In #{build_config_file}, topic group #{info[:group]}, one of the topics is not using a string for the '#{topic_key}' setting; current value is #{item[topic_key].inspect}"
```

Comment:
See above; if you are aware of a YAML parser that records source file line numbers, I'll use it.